### ISSUE TITLE
Inconsistent output resolution between gdalwarp and gdal_translate

### ISSUE THREAD

**Issue by mehran66** [OSGeo/gdal#3160]
## Expected behavior and actual behavior.

I am creating a COG file using both the gdalwarp and gdal_translate as follows but the resolution of the output is different:

gdal_translate block28_transparent_mosaic_group1.tif out_28.tif -of COG -co COMPRESS=JPEG -co NUM_THREADS=10

gdalwarp block28_transparent_mosaic_group1.tif out_28.tif -of COG -co COMPRESS=JPEG -co NUM_THREADS=10

The resolution of input image is 5 cm. The output of gdal_translate is 5 cm as well, but the gdalwarp leads to a resolution of 4.999.

## Operating system
Ubuntu 20.04 64 bit

## GDAL version and provenance
GDAL 3.1.4, released 2020/10/20

**Comment by rouault** [comment]
Please provide the input image, or at list the output of ``gdalinfo block28_transparent_mosaic_group1.tif ``

**Comment by mehran66** [comment]
I cannot share that data...but I tested this on another random file on openaerialmap and see the same issue, though a very small difference in the output resolution.
https://map.openaerialmap.org/#/-18.6328125,18.562947442888312,3/latest/5fa8540eaa6fdb00059cc1aa?_k=mwuwjg

**Comment by jratike80** [comment]
If the resolution is critical I suggest to define it explicitly with the target resolution option `-tr` that both utilities support. The resolution goes hand in hand with the image size and one pixel change in width or height leads to slight change in resolution. In your case the image size and resolution should remain the same but with `-tr` you can take the control.

**Comment by mehran66** [comment]
No sure why the link does not work anymore. It was just a random image from the openaerialmap.
Here is the gdalinfo for the block28_transparent_mosaic_group1.tif:
```
Driver: GTiff/GeoTIFF
Files: /home/mehran/backup/code/gis/imagery2grid/input-imagery/2019/block28_transparent_mosaic_group1.tif
Size is 49907, 54426
Coordinate System is:
PROJCRS["WGS 84 / UTM zone 35S",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["UTM zone 35S",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",27,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",500000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",10000000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["unknown"],
        AREA["World - S hemisphere - 24°E to 30°E - by country"],
        BBOX[-80,24,0,30]],
    ID["EPSG",32735]]
Data axis to CRS axis mapping: 1,2
Origin = (639737.531420000013895,8300646.355050000362098)
Pixel Size = (0.050000000000000,-0.050000000000000)
Metadata:
  AREA_OR_POINT=Area
  TIFFTAG_SOFTWARE=pix4dmapper
Image Structure Metadata:
  COMPRESSION=LZW
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  (  639737.531, 8300646.355) ( 28d18' 7.00"E, 15d22' 1.67"S)
Lower Left  (  639737.531, 8297925.055) ( 28d18' 7.55"E, 15d23'30.22"S)
Upper Right (  642232.881, 8300646.355) ( 28d19'30.68"E, 15d22' 1.17"S)
Lower Right (  642232.881, 8297925.055) ( 28d19'31.24"E, 15d23'29.72"S)
Center      (  640985.206, 8299285.705) ( 28d18'49.12"E, 15d22'45.70"S)
Band 1 Block=49907x1 Type=Byte, ColorInterp=Red
  Overviews: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
  Mask Flags: PER_DATASET ALPHA 
  Overviews of mask band: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
Band 2 Block=49907x1 Type=Byte, ColorInterp=Green
  Overviews: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
  Mask Flags: PER_DATASET ALPHA 
  Overviews of mask band: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
Band 3 Block=49907x1 Type=Byte, ColorInterp=Blue
  Overviews: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
  Mask Flags: PER_DATASET ALPHA 
  Overviews of mask band: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
Band 4 Block=49907x1 Type=Byte, ColorInterp=Alpha
  Overviews: 24954x27213, 12477x13607, 6239x6804, 3120x3402, 1560x1701, 780x851, 390x426, 195x213, 98x107
```

**Comment by jratike80** [comment]
I could not re-produce with these commands:

```
gdal_create -of GTiff -bands 4 -co tiled=yes -co compress=lzw -co bigtiff=yes -outsize 49907 54426 -a_srs epsg:32735 -a_ullr 639737.531 8300646.355 642232.881 8297925.055 resolution_test.tif
gdal_translate resolution_test.tif resolution_gdal_translate.tif -of COG -co COMPRESS=LZW
gdalwarp resolution_test.tif resolution_gdalwarp.tif -of COG -co COMPRESS=LZW
```

I tried gdalwarp also with JPEG compression.

Look at these lines in the gdalwarp report

```
Size is 49907, 54426
Pixel Size = (0.050000000000008,-0.050000000000008)
Upper Left  (  639737.531, 8300646.355) ( 28d18' 7.00"E, 15d22' 1.67"S)
Lower Right (  642232.881, 8297925.055) ( 28d19'31.24"E, 15d23'29.72"S)
```

Size, resolution, and extent goes hand in hand. So please check your data again and report not only resolution but image size and upper left - lower right coordinates as well.

Gdalwarp should not change any of those three measures if there is no re-projection. However, that is sort of special use case for warping and perhaps it is changing the size of the output by one pixel sometimes for some reason.

**Comment by mehran66** [comment]
I ran this process another time for a random image from openaerialmap. I am not sure that why when I share the url to the openaerialmap imagey, it does not open. So here is both the url and WMTS:
https://map.openaerialmap.org/#/13.7109375,42.94033923363181,2/latest/5fb25a434843720006bfdf62?_k=q6zlhr
https://tiles.openaerialmap.org/5fb259a94843720006bfdf60/0/5fb259a94843720006bfdf61/wmts

This image have a WGS 84 coordinate system. I ran both the following and you can see that the exported resolution is very different:

```
gdal_translate 5fb259a94843720006bfdf61.tif out_translat -of COG -co COMPRESS=JPEG -co NUM_THREADS=10
gdalinfo out_translat

Driver: GTiff/GeoTIFF
Files: out_translat
Size is 15910, 16414
Coordinate System is:
GEOGCRS["WGS 84",
    DATUM["World Geodetic System 1984",
        ELLIPSOID["WGS 84",6378137,298.257223563,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["geodetic latitude (Lat)",north,
            ORDER[1],
            ANGLEUNIT["degree",0.0174532925199433]],
        AXIS["geodetic longitude (Lon)",east,
            ORDER[2],
            ANGLEUNIT["degree",0.0174532925199433]],
    ID["EPSG",4326]]
Data axis to CRS axis mapping: 2,1
Origin = (37.470898233646203,56.948052989687497)
Pixel Size = (0.000000821492000,-0.000000448997000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  COMPRESSION=YCbCr JPEG
  INTERLEAVE=PIXEL
  LAYOUT=COG
  SOURCE_COLOR_SPACE=YCbCr
Corner Coordinates:
Upper Left  (  37.4708982,  56.9480530) ( 37d28'15.23"E, 56d56'52.99"N)
Lower Left  (  37.4708982,  56.9406832) ( 37d28'15.23"E, 56d56'26.46"N)
Upper Right (  37.4839682,  56.9480530) ( 37d29' 2.29"E, 56d56'52.99"N)
Lower Right (  37.4839682,  56.9406832) ( 37d29' 2.29"E, 56d56'26.46"N)
Center      (  37.4774332,  56.9443681) ( 37d28'38.76"E, 56d56'39.73"N)
Band 1 Block=512x512 Type=Byte, ColorInterp=Red
  Min=0.000 Max=255.000 
  Minimum=0.000, Maximum=255.000, Mean=183.431, StdDev=59.603
  Overviews: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Mask Flags: PER_DATASET 
  Overviews of mask band: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Unit Type: metre
  Metadata:
    STATISTICS_MAXIMUM=255
    STATISTICS_MEAN=183.43075676916
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=59.602701890689
Band 2 Block=512x512 Type=Byte, ColorInterp=Green
  Min=0.000 Max=255.000 
  Minimum=0.000, Maximum=255.000, Mean=182.836, StdDev=59.658
  Overviews: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Mask Flags: PER_DATASET 
  Overviews of mask band: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Unit Type: metre
  Metadata:
    STATISTICS_MAXIMUM=255
    STATISTICS_MEAN=182.8361499707
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=59.65845968101
Band 3 Block=512x512 Type=Byte, ColorInterp=Blue
  Min=0.000 Max=255.000 
  Minimum=0.000, Maximum=255.000, Mean=175.606, StdDev=65.585
  Overviews: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Mask Flags: PER_DATASET 
  Overviews of mask band: 7955x8207, 3978x4104, 1989x2052, 995x1026, 498x513, 249x257
  Unit Type: metre
  Metadata:
    STATISTICS_MAXIMUM=255
    STATISTICS_MEAN=175.6057986326
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=65.584745376665

```

```
gdalwarp 5fb259a94843720006bfdf61.tif out_warp -of COG -co COMPRESS=JPEG -co NUM_THREADS=10
gdalinfo out_warp

Driver: GTiff/GeoTIFF
Files: out_warp
Size is 19912, 11228
Coordinate System is:
GEOGCRS["WGS 84",
    DATUM["World Geodetic System 1984",
        ELLIPSOID["WGS 84",6378137,298.257223563,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["geodetic latitude (Lat)",north,
            ORDER[1],
            ANGLEUNIT["degree",0.0174532925199433]],
        AXIS["geodetic longitude (Lon)",east,
            ORDER[2],
            ANGLEUNIT["degree",0.0174532925199433]],
    ID["EPSG",4326]]
Data axis to CRS axis mapping: 2,1
Origin = (37.470898233646203,56.948052989687497)
Pixel Size = (0.000000656388994,-0.000000656388994)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  COMPRESSION=YCbCr JPEG
  INTERLEAVE=PIXEL
  LAYOUT=COG
  SOURCE_COLOR_SPACE=YCbCr
Corner Coordinates:
Upper Left  (  37.4708982,  56.9480530) ( 37d28'15.23"E, 56d56'52.99"N)
Lower Left  (  37.4708982,  56.9406831) ( 37d28'15.23"E, 56d56'26.46"N)
Upper Right (  37.4839683,  56.9480530) ( 37d29' 2.29"E, 56d56'52.99"N)
Lower Right (  37.4839683,  56.9406831) ( 37d29' 2.29"E, 56d56'26.46"N)
Center      (  37.4774332,  56.9443680) ( 37d28'38.76"E, 56d56'39.72"N)
Band 1 Block=512x512 Type=Byte, ColorInterp=Red
  Overviews: 9956x5614, 4979x2807, 2489x1404, 1245x702, 623x351, 312x176
  Unit Type: metre
Band 2 Block=512x512 Type=Byte, ColorInterp=Green
  Overviews: 9956x5614, 4979x2807, 2489x1404, 1245x702, 623x351, 312x176
  Unit Type: metre
Band 3 Block=512x512 Type=Byte, ColorInterp=Blue
  Overviews: 9956x5614, 4979x2807, 2489x1404, 1245x702, 623x351, 312x176
  Unit Type: metre

```

**Comment by jratike80** [comment]
That behavior is unrelated. Gdalwarp likes to create square pixels with width=height. 

```
Size is 15910, 16414
Pixel Size = (0.000000821492000,-0.000000448997000)
Converted
Size is 19912, 11228
Pixel Size = (0.000000656388994,-0.000000656388994)
```

Gdalwarp is mainly a warping tool and it is generally not guaranteed to maintain the original resolution. When it warps it is usually even impossible. If you want to output to have some exact pixel size, just use the `-tr` option. Or if you want exact size in pixels, use ´-ts`.

There is probably something to improve in the documentation, perhaps the right place is this wiki document https://trac.osgeo.org/gdal/wiki/UserDocs/GdalWarp.

**Comment by rouault** [comment]
Fixed per documentation addition in 18dc512
